docs/proposals/05-etcd-operator-tasks.md

```diff
@@ -0,0 +1,396 @@
+
+In this proposal, the authors recommend the implementation of a single controller with the ability to orchestrate these "out-of-band" operator tasks. The controller would initiate actions based on the `taskType` specified in the spec(desired state) of the EtcdOperatorTasks custom resource.
+
+### Custom Resource API
+
+The authors propose to add the `EtcdOperatorTasks` custom resource API to etcd-druid APIs and initially introduce it with `v1alpha1` version.
+
+#### Spec
+
+The authors propose that the following fields should be specified in the spec (desired state) of the EtcdOperatorTasks custom resource.
+
+- To capture the type of "out-of-band" operator task to be performed, `taskType` field is defined in spec. It can have values of all supported "out-of-band" operations/tasks eg. "OnDemandSnaphotTask", "QuorumLossRecoveryTask" etc.
+- To capture the configuration specific to each task, a `taskConfig` field is defined in spec as type [RawExtension](https://github.com/kubernetes/apimachinery/blob/829ed199f4e0454344a5bc5ef7859a01ef9b8e22/pkg/runtime/types.go#L49-L102). The controller (for instance, the EtcdOperatorTasks Controller) needs to parse this data, which comes as a RawExtension, according to the schema defined for each task.
+- `.spec.ttlSecondsAfterFinished` defines a `Time-to-live` to garbage collect the custom resource after completion of task.
+
+```go
+// EtcdOperatorTasksSpec is the spec for a EtcdOperatorTasks resource.
+type EtcdOperatorTasksSpec struct {
+
+  // TaskType is a field that specifies the type of out-of-band operator task to be performed. 
+  // It can have values of all supported operation/task type eg. "OnDemandSnapshotTask" etc.
+  TaskType *string `json:"taskType"`
+
+  // TTLSecondsAfterFinished define time-to-libe to garbage collect the custom resource after completion of task. 
+  TTLSecondsAfterFinished  int32 `json:"ttlSecondsAfterFinished"`
+
+  // TaskConfig is a task specific configuration.
+  TaskConfig *runtime.RawExtension `json:"taskConfig,omitempty"`
+}
+```
+
+#### Status
+
+The authors propose that the following fields should be specified in the status (current state) of the EtcdOperatorTasks custom resource.
+
+- To capture the Status of the task a `taskStatus` field is define in status.
+- If operation involves many stages, so to capture the status of intermediate or any stage, `.status.lastOperations` will be useful.
+- `.status.lastError` to capture the last error occured for the task.
+
+```go
+
+// EtcdOperatorTasksStatus is the status for a EtcdOperatorTasks resource.
+type EtcdOperatorTasksStatus struct {
+
+  // ObservedGeneration is the most recent generation observed for the resource.
+  ObservedGeneration *int64 `json:"observedGeneration,omitempty"`
+  
+  // Status of the last operation, one of pending, in-progress, completed, rejected or failed.
+  TaskStatus TaskState `json:"taskStatus"`
```

Comment:
Rename this to `State` - still not clear if this is really required. Need to revisit this field.